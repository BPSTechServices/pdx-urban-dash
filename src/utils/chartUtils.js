import React from 'react';
import LineChart from '../components/viz/LineChart';
import { BarChartStacked, BarChartClustered, BarChartGrouped } from '../components/viz/BarCharts/BarCharts';
import DonutChart from '../components/viz/DonutChart';

// eslint-disable-next-line import/prefer-default-export
export const genChart = (chartJson, colorScheme, className) => {
  if (chartJson === undefined) {
    return null;
  }
  // All charts use these elements
  const fInput = { title: chartJson.title, colorScheme, className };
  if (chartJson.chart_type === 'LINE') {
    const data = [];
    chartJson.data_sets.forEach((chartData) => {
      const values = [];
      chartData.data_values[0].map((e, i) => values.push({ x: e, y: chartData.data_values[1][i] }));
      data.push({
        dataSetName: chartData.title,
        lineColor: chartData.color,
        description: chartData.description,
        showTrendLine: chartData.show_trendLine,
        values,
      });
    });
    fInput.data = data;
    return <LineChart {...fInput} />;
  } if (chartJson.chart_type.startsWith('BAR')) {
    const data = [];
    chartJson.data_sets.forEach((chartData) => {
      const values = [];
      chartData.data_values[0].map((e, i) => values.push({ x: e, y: chartData.data_values[1][i] }));
      data.push({
        dataSetName: chartData.title,
        barColor: chartData.color,
        description: chartData.description,
        values,
      });
    });
    fInput.data = data;
    if (chartJson.chart_type === 'BARCLUSTERED' || chartJson.chart_type === 'BAR') {
      return <BarChartClustered {...fInput} />;
    }
    if (chartJson.chart_type === 'BARSTACKED') {
      return <BarChartStacked {...fInput} />;
    }
    if (chartJson.chart_type === 'BARGROUPED') {
      return <BarChartGrouped {...fInput} />;
    }
  } else if (chartJson.chart_type.startsWith('DONUT')) {
    const data = [];
    chartJson.data_sets.forEach((chartData) => {
      chartData.data_values[0].map((e, i) => data.push(
        {
          color: e,
          name: chartData.data_values[1][i],
          value: chartData.data_values[2][i],
        },
      ));
    });
    fInput.data = data;
    return <DonutChart {...fInput} />;
  }

  return null;
};
