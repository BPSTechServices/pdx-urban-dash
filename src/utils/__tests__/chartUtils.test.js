import React from 'react';
import each from 'jest-each';
import { shallow } from 'enzyme';
import { genChart } from '../chartUtils';

each([
  [undefined],
  [{
    targetTrend: 'UP',
    target: 100,
    title: 'dummy line graph',
    axisLabels: [
      'x',
      'y',
    ],
    type: 'LINE',
    dataSets: [
      {
        showTrendLine: true,
        description: 'first datapoint',
        title: 'line1',
        color: 'RED',
        dataValues: [
          [
            1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
          ],
          [
            2, 5, 1, 6, 8, 9, 3, 4, 7, 8,
          ],
        ],
        metaData: '',
      },
    ],
  }],
  [{
    metaData: 'metaData placeholder',
    title: 'bar1',
    description: 'Dummy bar chart 1: 7 bars of random positive size.',
    type: 'BAR',
    categories: [
      'category 4',
      'category 5',
      'category 3',
      'category 7',
      'category 2',
      'category 1',
      'category 6',
    ],
    target: 30.0,
    targetTrend: 'STABLE',
    axisLabels: [
      [
        'bar_labels',
        'bar values',
      ],
    ],
    dataSets: [
      {
        metaData: 'metaData placeholder.',
        title: "bar1 dataset member 1. Axis label(s): ['bar_labels', 'bar values']",
        description: 'Contains bar1 data for type BAR chart.',
        color: 'YELLOW',
        showTrendLine: true,
        dataValues: [
          [
            'bar 1',
            'bar 2',
            'bar 3',
            'bar 4',
            'bar 5',
            'bar 6',
            'bar 7',
            'bar 8',
          ],
          [
            78,
            35,
            72,
            99,
            33,
            78,
            75,
            91,
          ],
        ],
      },
    ],
  },
  {
    targetTrend: 'UP',
    target: 100,
    title: 'dummy line graph',
    axisLabels: [
      'x',
      'y',
    ],
    type: 'DONUT',
    dataSets: [
      {
        showTrendLine: true,
        description: 'first datapoint',
        title: 'line1',
        color: 'RED',
        dataValues: [
          [
            'RED',
            'BLUE',
          ],
          [
            'derp',
            'foo',
          ],
          [
            30,
            70,
          ],
        ],
        metaData: '',
      },
    ],
  },
  ]]).test('genChart with obj: %o', (obj) => {
  if (obj === undefined) {
    expect(genChart(obj)).toBeNull();
  } else {
    const chart = <genChart chartJson={obj} colorScheme="light" />;
    const wrapper = shallow(<chart />);
    expect(wrapper).toMatchSnapshot();
  }
});
