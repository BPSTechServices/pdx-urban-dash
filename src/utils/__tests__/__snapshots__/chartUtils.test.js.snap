// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genChart with obj: { metaData: 'metaData placeholder',
  title: 'bar1',
  description: 'Dummy bar chart 1: 7 bars of random positive size.',
  type: 'BAR',
  categories:
   [ 'category 4',
     'category 5',
     'category 3',
     'category 7',
     'category 2',
     'category 1',
     'category 6',
     [length]: 7 ],
  target: 30,
  targetTrend: 'STABLE',
  axisLabels: [ [ 'bar_labels', 'bar values', [length]: 2 ], [length]: 1 ],
  dataSets:
   [ { metaData: 'metaData placeholder.',
       title:
        'bar1 dataset member 1. Axis label(s): [\\'bar_labels\\', \\'bar values\\']',
       description: 'Contains bar1 data for type BAR chart.',
       color: 'YELLOW',
       showTrendLine: true,
       dataValues:
        [ [ 'bar 1',
            'bar 2',
            'bar 3',
            'bar 4',
            'bar 5',
            'bar 6',
            'bar 7',
            'bar 8',
            [length]: 8 ],
          [ 78, 35, 72, 99, 33, 78, 75, 91, [length]: 8 ],
          [length]: 2 ] },
     [length]: 1 ] } 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <chart />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": "chart",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": "chart",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`genChart with obj: { targetTrend: 'UP',
  target: 100,
  title: 'dummy line graph',
  axisLabels: [ 'x', 'y', [length]: 2 ],
  type: 'LINE',
  dataSets:
   [ { showTrendLine: true,
       description: 'first datapoint',
       title: 'line1',
       color: 'RED',
       dataValues:
        [ [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, [length]: 10 ],
          [ 2, 5, 1, 6, 8, 9, 3, 4, 7, 8, [length]: 10 ],
          [length]: 2 ],
       metaData: '' },
     [length]: 1 ] } 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <chart />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": "chart",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": "chart",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
